File: app/src/main/java/com/prey/PreyScheduled.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2015 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import java.util.Calendar;

import com.prey.receivers.AlarmScheduledReceiver;

public class PreyScheduled {

    private static PreyScheduled instance = null;
    private Context context = null;
    private AlarmManager alarmMgr = null;
    private PendingIntent alarmIntent = null;
    private PreyScheduled(Context context) {
        this.context = context;
    }

    public synchronized static PreyScheduled getInstance(Context context) {
        if (instance == null) {
            instance = new PreyScheduled(context);
        }
        return instance;
    }

    public void run(int minute) {
        final Context ctx = context;
        if (minute > 0) {
            reset();
            Intent intent = new Intent(context, AlarmScheduledReceiver.class);
            alarmIntent = PendingIntent.getBroadcast(context, 0, intent, 0);
            alarmMgr = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
            Calendar calendar = Calendar.getInstance();
            calendar.setTimeInMillis(System.currentTimeMillis());
            calendar.add(Calendar.MINUTE,minute);
            if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.KITKAT) {
                alarmMgr.setRepeating(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), 1000 * 60 * minute, alarmIntent);
            } else {
                alarmMgr.setInexactRepeating(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), 1000 * 60 * minute, alarmIntent);
            }
            PreyLogger.d("SCHEDULE_____________start scheduled [" + minute + "] alarmIntent");
        }
    }

    public void reset() {
        if (alarmMgr != null) {
            PreyLogger.d("_________________shutdown scheduled alarmIntent");
            alarmMgr.cancel(alarmIntent);
        }
    }

}

File: app/src/main/java/com/prey/PreyScheduled.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build

import com.prey.receivers.AlarmScheduledReceiver

/**
 * A utility class for scheduling and canceling alarms.
 */
class PreyScheduled {

    private lateinit var alarmManager: AlarmManager
    private lateinit var pendingIntent: PendingIntent

    /**
     * Starts the alarm scheduling process.
     *
     * @param context The application context
     */
    fun start(context: Context) {
        val intent = Intent(context, AlarmScheduledReceiver::class.java)
        pendingIntent =
            PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_MUTABLE)
        alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        scheduleAlarm()
    }

    /**
     * Schedules an alarm to trigger at a specified interval.
     */
    private fun scheduleAlarm() {
        val intervalMinutes = 15
        val triggerTime = System.currentTimeMillis()
        val interval = (1000 * 60 * intervalMinutes).toLong()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, triggerTime, interval, pendingIntent)
        } else {
            alarmManager.setInexactRepeating(
                AlarmManager.RTC_WAKEUP,
                triggerTime,
                interval,
                pendingIntent
            )
        }
    }

    /**
     * Cancels any currently scheduled alarms.
     */
    fun cancel() {
        try {
            alarmManager.cancel(pendingIntent)
        } catch (e: Exception) {
            PreyLogger.d("----------Error PreyScheduled :${e.message}")
        }
    }

    companion object {
        private var instance: PreyScheduled? = null
        fun getInstance(): PreyScheduled =
            instance ?: PreyScheduled().also { instance = it }
    }
}